src/ppx/geometry.cpp

```diff
@@ -19,6 +19,334 @@
+    {
+        return pGeom->mVertexBuffers[bufferIndex].GetElementSize();
+    }
+
+    uint32_t GetVertexBindingAttributeCount(const Geometry* pGeom, uint32_t bindingIndex) const
+    {
+        return pGeom->mCreateInfo.vertexBindings[bindingIndex].GetAttributeCount();
+    }
+
+    uint32_t GetVertexBindingStride(const Geometry* pGeom, uint32_t bindingIndex) const
+    {
+        return pGeom->mCreateInfo.vertexBindings[bindingIndex].GetStride();
+    }
+
+    uint32_t GetVertexBindingCount(const Geometry* pGeom) const
+    {
+        return pGeom->mCreateInfo.vertexBindingCount;
+    }
+
+    grfx::VertexSemantic GetVertexBindingAttributeSematic(const Geometry* pGeom, uint32_t bindingIndex, uint32_t attrIndex) const
+    {
+        const grfx::VertexAttribute* pAttribute = nullptr;
+        Result                       ppxres     = pGeom->mCreateInfo.vertexBindings[bindingIndex].GetAttribute(attrIndex, &pAttribute);
+        PPX_ASSERT_MSG((ppxres == ppx::SUCCESS), "attribute not found at index=" << attrIndex);
+        return pAttribute->semantic;
+    }
+
+    uint32_t GetPositionBufferIndex(const Geometry* pGeom) const { return pGeom->mPositionBufferIndex; }
+    uint32_t GetNormalBufferIndex(const Geometry* pGeom) const { return pGeom->mNormaBufferIndex; }
+    uint32_t GetColorBufferIndex(const Geometry* pGeom) const { return pGeom->mColorBufferIndex; }
+    uint32_t GetTexCoordBufferIndex(const Geometry* pGeom) const { return pGeom->mTexCoordBufferIndex; }
+    uint32_t GetTangentBufferIndex(const Geometry* pGeom) const { return pGeom->mTangentBufferIndex; }
+    uint32_t GetBitangentBufferIndex(const Geometry* pGeom) const { return pGeom->mBitangentBufferIndex; }
+
+    void SetPositionBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mPositionBufferIndex = index; }
+    void SetNormalBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mNormaBufferIndex = index; }
+    void SetColorBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mColorBufferIndex = index; }
+    void SetTexCoordBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mTexCoordBufferIndex = index; }
+    void SetTangentBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mTangentBufferIndex = index; }
+    void SetBitangentBufferIndex(Geometry* pGeom, uint32_t index) { pGeom->mBitangentBufferIndex = index; }
+};
+
+class VertexDataProcessorPlanar : public IVertexDataProcessor
+{
+public:
+    virtual bool Validate(Geometry* pGeom) override
+    {
+        const uint32_t vertexBindingCount = GetVertexBindingCount(pGeom);
+        for (uint32_t i = 0; i < vertexBindingCount; ++i) {
+            if (GetVertexBindingAttributeCount(pGeom, i) != 1) {
+                PPX_ASSERT_MSG(false, "planar layout binding must have 1 attribute");
+                return false;
+            }
+        }
+        return true;
+    }
+    virtual Result UpdateVertexBuffer(Geometry* pGeom) override
```

Comment:
There should be an empty line between functions.

Reply:
fixed